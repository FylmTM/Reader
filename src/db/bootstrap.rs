use rusqlite::{Connection, NO_PARAMS};

const SCHEMA_VERSION: u16 = 1;
// language=SQLite
const SCHEMA_VERSION_QUERY: &str = "
create table if not exists schema_version
(
    id      integer primary key,
    version integer
);

insert or ignore into schema_version(id, version)
values (1, 0);
";

pub fn initialize_schema(conn: &Connection) {
    conn.execute_batch(SCHEMA_VERSION_QUERY)
        .expect("Failed to create schema_version table.");

    let mut current_version: u16 = conn
        .query_row(
            // language=SQLite
            "select version from schema_version",
            NO_PARAMS,
            |row| row.get(0),
        )
        .expect("Failed to retrieve current schema version.");

    let version_before = current_version;

    info!(
        "Current schema version=V{}, required=V{}.",
        current_version, SCHEMA_VERSION
    );

    if current_version == 0 {
        info!("Applying V1 migration.");
        conn.execute_batch(include_str!("migrations/V1_initial.sql"))
            .expect("Schema creation failed.");
        current_version = 1;
    }

    if version_before != current_version {
        info!("Set current schema version to V{}.", current_version);
        conn.execute(
            // language=SQLite
            "update schema_version set version = ? where id = 1",
            &[current_version],
        )
        .expect("Failed to update current schema version.");
    }
}

pub fn initialize_fixture(conn: &Connection) {
    // language=SQLite
    let query = "
        insert or ignore into users(id, username, api_key)
        values (1, 'me', 'api_key');
    ";
    conn.execute_batch(query).expect("Failed to apply fixture.");
}
