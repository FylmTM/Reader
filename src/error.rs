use std::io::Cursor;

use rocket::http::Status;
use rocket::response::Responder;
use rocket::{Request, Response};

use ApplicationError::*;

use crate::api::ApiError;

pub type Result<T> = std::result::Result<T, Error>;

#[derive(Debug)]
pub enum Error {
    App(ApplicationError),
    Pool(r2d2::Error),
    DB(rusqlite::Error),
    HttpClient(reqwest::Error),
    RSS(rss::Error),
    Atom(atom_syndication::Error),
    Chrono(chrono::ParseError),
}

#[derive(Debug)]
pub enum ApplicationError {
    AuthenticationApiKeyNotFound,
    FailedGetUrlContent(String),
    RSSParsingMissingItemLink,
    RSSParsingMissingItemTitle,
    AtomParsingMissingEntryLink,
    QueryEntityAlreadyExists,
    QueryEntityNotFound,
    QueryUnexpectedRowCount { expected: i64, actual: i64 },
}

impl Error {
    fn describe(self) -> (Status, String) {
        match self {
            Error::App(AuthenticationApiKeyNotFound) => {
                (Status::NotFound, "API key does not exists.".to_string())
            }
            Error::App(QueryUnexpectedRowCount { expected, actual }) => (
                Status::BadRequest,
                format!("Expected row count {}, got {}.", expected, actual),
            ),
            Error::DB(rusqlite::Error::QueryReturnedNoRows) => {
                (Status::NotFound, "Entity not found.".to_string())
            }
            _ => (Status::InternalServerError, format!("{:?}", self)),
        }
    }
}

impl<'r> Responder<'r> for Error {
    fn respond_to(self, _: &Request) -> std::result::Result<Response<'r>, Status> {
        let (http_status, message) = self.describe();

        let json = serde_json::to_string(&ApiError::with(message))
            .expect("Failed to serialize api error.");

        Response::build()
            .header(rocket::http::ContentType::JSON)
            .sized_body(Cursor::new(json))
            .status(http_status)
            .ok()
    }
}

impl From<rusqlite::Error> for Error {
    fn from(error: rusqlite::Error) -> Error {
        Error::DB(error)
    }
}

impl From<r2d2::Error> for Error {
    fn from(error: r2d2::Error) -> Error {
        Error::Pool(error)
    }
}

impl From<reqwest::Error> for Error {
    fn from(error: reqwest::Error) -> Error {
        Error::HttpClient(error)
    }
}

impl From<rss::Error> for Error {
    fn from(error: rss::Error) -> Error {
        Error::RSS(error)
    }
}

impl From<atom_syndication::Error> for Error {
    fn from(error: atom_syndication::Error) -> Error {
        Error::Atom(error)
    }
}

impl From<chrono::ParseError> for Error {
    fn from(error: chrono::ParseError) -> Error {
        Error::Chrono(error)
    }
}
