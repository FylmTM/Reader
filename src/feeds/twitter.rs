use rusqlite::Connection;
use std::thread;
use std::time::Duration;

pub fn update(_conn: &Connection) {
    thread::sleep(Duration::from_secs(5));
}
