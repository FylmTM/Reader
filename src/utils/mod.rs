use rocket::fairing::{Fairing, Info, Kind};
use rocket::{Request, Response};

pub struct AppHeaders();

impl Fairing for AppHeaders {
    fn info(&self) -> Info {
        Info {
            name: "Application Headers",
            kind: Kind::Response,
        }
    }

    fn on_response(&self, _request: &Request, response: &mut Response) {
        response.set_raw_header(
            "Content-Security-Policy",
            "default-src 'none'; script-src 'self' 'unsafe-inline'; connect-src 'self'; img-src *; style-src 'self' 'unsafe-inline' https://fonts.googleapis.com; font-src 'self' https://fonts.gstatic.com",
        );
    }
}
