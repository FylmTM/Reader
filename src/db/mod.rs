/// This whole module is using vanilla rusqlite, without Diesel.
/// At the time of creating adding database support author wanted to stay
/// as close as possible to low-level libraries and write plain-SQL queries.
///
/// Because of that you can find poor-mans migrations module here, and some
/// nasty looking string manipulations for dynamic queries.
///
/// Author might reconsider his choice in future.
pub use rusqlite::{Connection, NO_PARAMS};

pub use entities::*;
pub use pool::{Pool, PoolConnection};

use crate::error::{ApplicationError, Error, Result};
use std::collections::HashMap;

pub mod bootstrap;
pub mod entities;
pub mod pool;

const PAGE_SIZE: usize = 50;

pub trait Queries {
    fn find_user_by_api_key(&self, api_key: &str) -> Result<User>;
    fn find_user_ids_by_feed(&self, feed_id: FeedId) -> Result<Vec<UserId>>;
    fn find_categories_by_user(&self, user_id: UserId) -> Result<Vec<Category>>;
    fn find_feeds(&self) -> Result<Vec<Feed>>;
    fn find_feeds_by_category(&self, category_id: CategoryId) -> Result<Vec<Feed>>;
    fn find_feed(&self, id: FeedId) -> Result<Feed>;
    fn find_posts(
        &self,
        user_id: UserId,
        from_post_id: Option<PostId>,
        is_read_later: Option<bool>,
        is_read: Option<bool>,
        category_id: Option<CategoryId>,
        feed_id: Option<FeedId>,
    ) -> Result<Page<UserPost>>;
    fn find_post_content_by_user(
        &self,
        user_id: UserId,
        post_id: PostId,
    ) -> Result<UserPostContent>;
    fn set_post_is_read(&self, user_id: UserId, post_id: PostId, is_read: bool) -> Result<()>;
    fn set_post_is_read_later(
        &self,
        user_id: UserId,
        post_id: PostId,
        is_read_later: bool,
    ) -> Result<()>;
    fn count_posts_unread(&self, user_id: UserId) -> Result<UserPostsUnreadCount>;
    fn save_post(&self, feed_id: FeedId, post: &Post) -> Result<PostId>;
    fn save_user_post(&self, user_id: UserId, post_id: PostId) -> Result<UserPostId>;
}

impl Queries for Connection {
    fn find_user_by_api_key(self: &Connection, api_key: &str) -> Result<User> {
        // language=SQLite
        let query = "
            select u.id, u.username
            from users u
            where u.api_key = :api_key
        ";
        let user = self.query_row_named(query, &[(":api_key", &api_key)], |row| {
            Ok(User {
                id: row.get("id")?,
                username: row.get("username")?,
            })
        })?;
        Ok(user)
    }

    fn find_user_ids_by_feed(self: &Connection, feed_id: FeedId) -> Result<Vec<i64>> {
        // language=SQLite
        let query = "
            select u.id
            from users u
            inner join user_categories uc on u.id = uc.user_id
            inner join user_category_feeds ucf on uc.id = ucf.category_id
            where ucf.feed_id = :feed_id
        ";
        let mut statement = self.prepare(query)?;
        let user_ids_rows =
            statement.query_map_named(&[(":feed_id", &feed_id)], |row| Ok(row.get("id")?))?;

        let mut user_ids = Vec::new();
        for user_id in user_ids_rows {
            user_ids.push(user_id?);
        }
        Ok(user_ids)
    }

    fn find_categories_by_user(self: &Connection, user_id: UserId) -> Result<Vec<Category>> {
        // language=SQLite
        let query = "
            select uc.id, uc.title
            from user_categories uc
            where uc.user_id = :user_id
        ";
        let mut statement = self.prepare(query)?;
        let categories_rows = statement.query_map_named(&[(":user_id", &user_id)], |row| {
            Ok(Category {
                id: row.get("id")?,
                title: row.get("title")?,
            })
        })?;

        let mut categories = Vec::new();
        for category in categories_rows {
            categories.push(category?)
        }
        Ok(categories)
    }

    fn find_feeds(self: &Connection) -> Result<Vec<Feed>> {
        // language=SQLite
        let query = "
            select f.id, f.kind, f.title, f.link, f.feed
            from feeds f
        ";
        let mut statement = self.prepare(query)?;
        let feeds_rows = statement.query_map(NO_PARAMS, map_row_to_feed)?;

        let mut feeds = Vec::new();
        for feed in feeds_rows {
            feeds.push(feed?)
        }
        Ok(feeds)
    }

    fn find_feeds_by_category(self: &Connection, category_id: CategoryId) -> Result<Vec<Feed>> {
        // language=SQLite
        let query = "
            select f.id, f.kind, ucf.title, f.link, f.feed
            from user_category_feeds ucf
            inner join feeds f on ucf.feed_id = f.id
            where ucf.category_id = :category_id
        ";
        let mut statement = self.prepare(query)?;
        let feeds_rows =
            statement.query_map_named(&[(":category_id", &category_id)], map_row_to_feed)?;

        let mut feeds = Vec::new();
        for feed in feeds_rows {
            feeds.push(feed?)
        }
        Ok(feeds)
    }

    //noinspection DuplicatedCode
    fn find_feed(self: &Connection, id: FeedId) -> Result<Feed> {
        // language=SQLite
        let query = "
            select f.id, f.kind, f.title, f.link, f.feed
            from feeds f
            where f.id = :id
        ";
        let feed = self.query_row_named(query, &[(":id", &id)], map_row_to_feed)?;
        Ok(feed)
    }

    fn find_posts(
        &self,
        user_id: UserId,
        from_post_id: Option<PostId>,
        is_read_later: Option<bool>,
        is_read: Option<bool>,
        category_id: Option<CategoryId>,
        feed_id: Option<FeedId>,
    ) -> Result<Page<UserPost>> {
        let sql_from_post_id = from_post_id
            .map(|_| "and up.post_id < :post_id")
            .unwrap_or("");
        let sql_is_read_later = is_read_later
            .map(|_| "and up.is_read_later = :is_read_later")
            .unwrap_or("");
        let sql_is_read = is_read.map(|_| "and up.is_read = :is_read").unwrap_or("");
        let sql_category_id = category_id
            .map(|_| "and ucf.category_id = :category_id")
            .unwrap_or("");
        let sql_feed_id = feed_id.map(|_| "and ucf.feed_id = :feed_id").unwrap_or("");

        // language=SQLite
        let query = format!(
            "
            select p.id, ucf.category_id, ucf.feed_id as feed_id, ucf.title as feed_title,
                   up.is_read, up.read_date, up.is_read_later,
                   p.link, p.title, p.date, p.summary, p.media_type, p.media_link, p.comments_link
            from user_posts up
            inner join posts p on up.post_id = p.id
            inner join user_category_feeds ucf on p.feed_id = ucf.feed_id
            where up.user_id = :user_id
            {}
            {}
            {}
            {}
            {}
            order by up.post_id desc
            limit {};
        ",
            sql_from_post_id,
            sql_is_read_later,
            sql_is_read,
            sql_category_id,
            sql_feed_id,
            PAGE_SIZE + 1
        );
        let mut statement = self.prepare(&query)?;

        let mut params: Vec<(&str, &dyn rusqlite::ToSql)> = Vec::new();
        params.push((":user_id", &user_id));

        if from_post_id.is_some() {
            params.push((":post_id", from_post_id.as_ref().unwrap()));
        }
        if is_read_later.is_some() {
            params.push((":is_read_later", is_read_later.as_ref().unwrap()));
        }
        if is_read.is_some() {
            params.push((":is_read", is_read.as_ref().unwrap()));
        }
        if category_id.is_some() {
            params.push((":category_id", category_id.as_ref().unwrap()));
        }
        if feed_id.is_some() {
            params.push((":feed_id", feed_id.as_ref().unwrap()));
        }

        let user_post_rows = statement.query_map_named(params.as_slice(), map_row_to_user_post)?;
        let mut user_posts = Vec::new();
        for user_post in user_post_rows {
            user_posts.push(user_post?)
        }

        // We are querying for PAGE_SIZE + 1.
        // If there is more elements than PAGE_SIZE, then there is another non-empty page.
        let has_next_page = user_posts.len() > PAGE_SIZE;
        user_posts.truncate(PAGE_SIZE);
        Ok(Page {
            has_next_page,
            items: user_posts,
        })
    }

    fn find_post_content_by_user(
        &self,
        user_id: UserId,
        post_id: PostId,
    ) -> Result<UserPostContent> {
        // language=SQLite
        let query = "
            select p.id, p.content
            from user_posts up
            inner join posts p on up.post_id = p.id
            where up.post_id = :post_id
            and up.user_id = :user_id
        ";
        let mut statement = self.prepare(query)?;
        let user_post_content = statement.query_row_named(
            &[(":user_id", &user_id), (":post_id", &post_id)],
            |row| {
                Ok(UserPostContent {
                    id: row.get("id")?,
                    content: row.get("content")?,
                })
            },
        )?;

        Ok(user_post_content)
    }

    fn set_post_is_read(&self, user_id: UserId, post_id: PostId, is_read: bool) -> Result<()> {
        // language=SQLite
        let query = "
            update user_posts
            set is_read = :is_read,
                read_date = :read_date
            where post_id = :post_id
            and user_id = :user_id
        ";
        let read_date: Option<chrono::DateTime<chrono::Utc>> = if is_read {
            Some(chrono::DateTime::from(chrono::Utc::now()))
        } else {
            None
        };
        let count = self.execute_named(
            query,
            &[
                (":user_id", &user_id),
                (":post_id", &post_id),
                (":is_read", &is_read),
                (":read_date", &read_date),
            ],
        )?;
        check_one_entity_updated(count)?;
        return Ok(());
    }

    fn set_post_is_read_later(
        &self,
        user_id: UserId,
        post_id: PostId,
        is_read_later: bool,
    ) -> Result<()> {
        // language=SQLite
        let query = "
            update user_posts
            set is_read_later = :is_read_later
            where post_id = :post_id
            and user_id = :user_id
        ";
        let count = self.execute_named(
            query,
            &[
                (":user_id", &user_id),
                (":post_id", &post_id),
                (":is_read_later", &is_read_later),
            ],
        )?;
        check_one_entity_updated(count)?;
        return Ok(());
    }

    fn count_posts_unread(&self, user_id: UserId) -> Result<UserPostsUnreadCount> {
        let mut count = UserPostsUnreadCount {
            all: 0,
            categories: HashMap::new(),
            feeds: HashMap::new(),
        };
        // language=SQLite
        let query = "
            select ucf.category_id as category_id, ucf.feed_id as feed_id, count(up.post_id) as count
            from user_posts up
            inner join posts p on up.post_id = p.id
            inner join user_category_feeds ucf on p.feed_id = ucf.feed_id
            where up.user_id = :user_id
            and up.is_read = false
            group by ucf.feed_id
        ";
        let mut statement = self.prepare(query)?;
        let rows = statement.query_map_named(&[(":user_id", &user_id)], |row| {
            Ok(CategoryFeedUnreadCount {
                category_id: row.get("category_id")?,
                feed_id: row.get("feed_id")?,
                count: row.get("count")?,
            })
        })?;

        for row in rows {
            let row = row?;
            count.all += row.count;
            count.categories.insert(
                row.category_id,
                count
                    .categories
                    .get(&row.category_id)
                    .map(|c| c + row.count)
                    .unwrap_or(row.count),
            );
            count.feeds.insert(row.feed_id, row.count);
        }

        Ok(count)
    }

    fn save_post(self: &Connection, feed_id: FeedId, post: &Post) -> Result<PostId> {
        // language=SQLite
        let query = "
            insert or ignore into posts(feed_id, unique_id, link, title, date, summary, content, media_type, media_link, comments_link)
            values (:feed_id, :unique_id, :link, :title, :date, :summary, :content, :media_type, :media_link, :comments_link)
        ";
        let count = self.execute_named(
            query,
            &[
                (":feed_id", &feed_id),
                (":unique_id", &post.unique_id),
                (":link", &post.link),
                (":title", &post.title),
                (":date", &post.date),
                (":summary", &post.summary),
                (":content", &post.content),
                (":media_type", &post.media_type),
                (":media_link", &post.media_link),
                (":comments_link", &post.comments_link),
            ],
        )?;
        check_one_entity_inserted(count)?;
        Ok(self.last_insert_rowid())
    }

    fn save_user_post(&self, user_id: UserId, post_id: PostId) -> Result<UserPostId> {
        // language=SQLite
        let query = "
            insert or ignore into user_posts(user_id, post_id)
            values (:user_id, :post_id)
        ";
        let count = self.execute_named(query, &[(":user_id", &user_id), (":post_id", &post_id)])?;
        check_one_entity_inserted(count)?;
        Ok(self.last_insert_rowid())
    }
}

fn map_row_to_user_post(row: &rusqlite::Row) -> rusqlite::Result<UserPost> {
    Ok(UserPost {
        id: row.get("id")?,
        category_id: row.get("category_id")?,
        feed_id: row.get("feed_id")?,
        feed_title: row.get("feed_title")?,
        is_read: row.get("is_read")?,
        read_date: row.get("read_date")?,
        is_read_later: row.get("is_read_later")?,
        link: row.get("link")?,
        title: row.get("title")?,
        date: row.get("date")?,
        summary: row.get("summary")?,
        media_type: row.get("media_type")?,
        media_link: row.get("media_link")?,
        comments_link: row.get("comments_link")?,
    })
}

fn map_row_to_feed(row: &rusqlite::Row) -> rusqlite::Result<Feed> {
    Ok(Feed {
        id: row.get("id")?,
        kind: row.get("kind")?,
        title: row.get("title")?,
        link: row.get("link")?,
        feed: row.get("feed")?,
    })
}

//noinspection DuplicatedCode
fn check_one_entity_inserted(count: usize) -> Result<()> {
    if count == 1 {
        Ok(())
    } else if count == 0 {
        Err(Error::App(ApplicationError::QueryEntityAlreadyExists))
    } else {
        Err(Error::App(ApplicationError::QueryUnexpectedRowCount {
            expected: 1,
            actual: count as i64,
        }))
    }
}

//noinspection DuplicatedCode
fn check_one_entity_updated(count: usize) -> Result<()> {
    if count == 1 {
        Ok(())
    } else if count == 0 {
        Err(Error::App(ApplicationError::QueryEntityNotFound))
    } else {
        Err(Error::App(ApplicationError::QueryUnexpectedRowCount {
            expected: 1,
            actual: count as i64,
        }))
    }
}
