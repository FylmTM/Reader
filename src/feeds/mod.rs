use std::thread;
use std::time::Duration;

use crate::db::{self, Queries};
use crate::error::{ApplicationError, Error, Result};

pub mod atom;
pub mod rss;
pub mod utils;

pub fn initialize(pool: &db::Pool, interval: u64) {
    let pool = pool.clone();

    info!("Schedule feeds update every {} minutes.", interval);
    thread::spawn(move || loop {
        thread::sleep(Duration::from_secs(5));
        loop {
            let mut conn = pool.get().expect("Failed to acquire db connection");
            update(&mut conn);

            info!("Next feed update in {} minutes.", interval);
            thread::sleep(Duration::from_secs(60 * interval));
        }
    });
}

pub fn update(mut conn: &mut db::Connection) {
    info!("Feeds update started.");
    match conn.find_feeds() {
        Ok(feeds) => {
            for feed in feeds {
                match update_feed(&mut conn, &feed) {
                    Ok(_) => debug!("{:?} feed updated.", feed),
                    Err(error) => error!("{:?} feed failed to update: {:?}", feed, error),
                };
            }
        }
        Err(error) => {
            error!("Failed to retrieve feeds: {:?}.", error);
        }
    }
    info!("Feeds update completed.");
}

pub fn update_feed(conn: &mut db::Connection, feed: &db::Feed) -> Result<()> {
    debug!("{:?} update started.", feed);

    let user_ids = conn.find_user_ids_by_feed(feed.id)?;
    if user_ids.is_empty() {
        debug!("{:?} does not belong to any user, skip update.", feed);
        return Ok(());
    }

    let posts = match feed.kind {
        db::FeedKind::RSS => rss::update(&feed),
        db::FeedKind::Atom => atom::update(&feed),
    }?;

    save_posts(conn, feed, &posts, &user_ids)?;

    Ok(())
}

fn save_posts(
    conn: &mut db::Connection,
    feed: &db::Feed,
    posts: &Vec<db::Post>,
    user_ids: &Vec<db::UserId>,
) -> Result<()> {
    for post in posts {
        let tx = conn.transaction()?;
        match tx.save_post(feed.id, &post) {
            Ok(post_id) => {
                debug!("{:?} saved.", post);
                for user_id in user_ids {
                    tx.save_user_post(*user_id, post_id)?;
                }
                tx.commit()?;
            }
            Err(Error::App(ApplicationError::QueryEntityAlreadyExists)) => {
                debug!("{:?} already exists.", &post)
            }
            Err(error) => return Err(error),
        }
    }
    Ok(())
}
