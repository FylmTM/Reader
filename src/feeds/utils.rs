use crate::error::{ApplicationError, Error, Result};

lazy_static! {
    static ref AMMONIA_TEXT: ammonia::Builder<'static> = {
        let mut builder = ammonia::Builder::default();
        builder.tags(std::collections::HashSet::new());
        builder
    };
}

pub fn get_url_content(link: &String) -> Result<String> {
    let mut response = reqwest::get(link)?;

    if response.status() != reqwest::StatusCode::OK {
        error!(
            "Failed to retrieve feed={} with status={}",
            link,
            response.status()
        );
        return Err(Error::App(ApplicationError::FailedGetUrlContent(
            link.clone(),
        )));
    }

    return Ok(response.text()?);
}

pub fn clean_to_text(html: &str) -> String {
    AMMONIA_TEXT.clean(html).to_string()
}

pub fn clean_to_safe_html(html: &str, base_url: &str) -> String {
    let mut builder = ammonia::Builder::default();
    builder.url_relative(ammonia::UrlRelative::RewriteWithBase(
        ammonia::Url::parse(base_url).expect("Failed to parse base url"),
    ));
    builder.clean(html).to_string()
}

pub fn find_image(html: &str) -> Option<String> {
    // language=RegExp
    let regexp = regex::Regex::new(r#"<img.*?src="(.+?)""#).unwrap();
    regexp
        .captures(html)
        .and_then(|captures| captures.get(1))
        .map(|src| src.as_str().to_string())
}
