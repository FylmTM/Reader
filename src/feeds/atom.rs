use crate::db::{Feed, MediaType, Post};
use crate::error::{ApplicationError, Error, Result};
use crate::feeds::utils::{clean_to_safe_html, clean_to_text, find_image, get_url_content};

pub fn update(feed: &Feed) -> Result<Vec<Post>> {
    let content = get_url_content(&feed.feed)?;
    let posts = parse(&content)?;
    Ok(posts)
}

fn parse(content: &str) -> Result<Vec<Post>> {
    let feed = content.parse::<atom_syndication::Feed>()?;
    let mut posts = Vec::with_capacity(feed.entries().len());

    for entry in feed.entries() {
        let unique_id = entry.id().to_string();
        let link = entry
            .links()
            .iter()
            .find(|link| link.rel() == "alternate")
            .or_else(|| entry.links().get(0))
            .map(|link| link.href())
            .ok_or(Error::App(ApplicationError::AtomParsingMissingEntryLink))?
            .to_string();
        let title = entry.title().to_string();
        let date = parse_date(entry.published().unwrap_or(entry.updated()))?;
        let summary = entry.summary().map(clean_to_text);
        let content = entry
            .content()
            .and_then(|content| content.value())
            .or(entry.summary())
            .map(|content| clean_to_safe_html(content, &link));
        let comments_link = None;

        let (media_type, media_link) = {
            content
                .as_ref()
                .and_then(|content| find_image(content))
                .map(|image| {
                    (
                        Some(MediaType {
                            mime: "image/*".to_string(),
                        }),
                        Some(image),
                    )
                })
                .unwrap_or((None, None))
        };

        let post = Post {
            id: -1, // Post is not retrieved from database
            unique_id,
            link,
            title,
            date,
            summary,
            content,
            media_type,
            media_link,
            comments_link,
        };
        posts.push(post);
    }

    // Items are sorted from newest to oldest, but for database insertion
    // we need them sorted from oldest to newest.
    posts.reverse();

    Ok(posts)
}

fn parse_date(date: &str) -> Result<chrono::DateTime<chrono::Utc>> {
    let result =
        chrono::DateTime::parse_from_rfc3339(date).map(|date| chrono::DateTime::from(date))?;

    Ok(result)
}

#[cfg(test)]
mod tests {
    use crate::feeds::atom::parse;

    #[test]
    fn test_parse_atom_github_rocket_releases() {
        let content = include_str!("../../static/mock/atom/github_rocket_releases.atom");
        let result = parse(&content).unwrap();
        assert_debug_snapshot!(result);
    }

    #[test]
    fn test_parse_atom_golang() {
        let content = include_str!("../../static/mock/atom/golang.atom");
        let result = parse(&content).unwrap();
        assert_debug_snapshot!(result);
    }

    #[test]
    fn test_parse_atom_julia_evans() {
        let content = include_str!("../../static/mock/atom/julia_evans.atom");
        let result = parse(&content).unwrap();
        assert_debug_snapshot!(result);
    }

    #[test]
    fn test_parse_atom_rust() {
        let content = include_str!("../../static/mock/atom/rust.atom");
        let result = parse(&content).unwrap();
        assert_debug_snapshot!(result);
    }
}
