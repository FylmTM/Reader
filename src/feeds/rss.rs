use crate::db::{Feed, MediaType, Post};
use crate::error::{ApplicationError, Error, Result};
use crate::feeds::utils::{clean_to_safe_html, clean_to_text, find_image, get_url_content};

pub fn update(feed: &Feed) -> Result<Vec<Post>> {
    let content = get_url_content(&feed.feed)?;
    let posts = parse(&content)?;

    Ok(posts)
}

fn parse(content: &str) -> Result<Vec<Post>> {
    let channel = content.parse::<rss::Channel>()?;
    let mut posts = Vec::with_capacity(channel.items().len());

    for item in channel.items() {
        let link = item
            .link()
            .ok_or(Error::App(ApplicationError::RSSParsingMissingItemLink))?
            .to_string();
        let unique_id = item
            .guid()
            .map(|guid| guid.value().to_string())
            .unwrap_or_else(|| link.clone());
        let title = item
            .title()
            .ok_or(Error::App(ApplicationError::RSSParsingMissingItemTitle))?
            .to_string();
        let date = match item.pub_date() {
            Some(date) => parse_date(date)?,
            None => chrono::DateTime::from(chrono::Utc::now()),
        };

        let summary = item.description().map(clean_to_text);
        let content = item
            .content()
            .or(item.description())
            .map(|content| clean_to_safe_html(content, &link));
        let comments_link = item.comments().map(String::from);

        let (media_type, media_link) = {
            if item.enclosure().is_some() {
                let enclosure = item.enclosure().unwrap();
                let media_type = {
                    let mime_type = enclosure.mime_type();
                    if mime_type.is_empty() {
                        None
                    } else if mime_type.parse::<mime::Mime>().is_err() {
                        None
                    } else {
                        Some(MediaType {
                            mime: mime_type.to_string(),
                        })
                    }
                };
                let media_link = {
                    let url = enclosure.url();
                    if url.is_empty() {
                        None
                    } else {
                        Some(url.to_string())
                    }
                };

                (media_type, media_link)
            } else if content.is_some() {
                // If there is no enclosure in feed, let's try to cut out first image from content
                let content = content.as_ref().unwrap();
                find_image(content)
                    .map(|image| {
                        (
                            Some(MediaType {
                                mime: "image/*".to_string(),
                            }),
                            Some(image),
                        )
                    })
                    .unwrap_or((None, None))
            } else {
                (None, None)
            }
        };

        let post = Post {
            id: -1, // Post is not retrieved from database
            unique_id,
            link,
            title,
            date,
            summary,
            content,
            media_type,
            media_link,
            comments_link,
        };
        posts.push(post);
    }

    // Items are sorted from newest to oldest, but for database insertion
    // we need them sorted from oldest to newest.
    posts.reverse();
    Ok(posts)
}

fn parse_date(date: &str) -> Result<chrono::DateTime<chrono::Utc>> {
    let result = chrono::DateTime::parse_from_rfc2822(date)
        .or_else(|_| {
            // chrono consider "-0000" timezone to be "missing"
            // We will blindly replace it with +0000 if it exists, and try to parse again.
            let date = date.replace("-0000", "+0000");
            chrono::DateTime::parse_from_rfc2822(&date)
        })
        .map(|date| chrono::DateTime::from(date))?;
    Ok(result)
}

#[cfg(test)]
mod tests {
    use crate::feeds::rss::{parse, parse_date};

    #[test]
    fn test_parse_rss_hacker_news() {
        let content = include_str!("../../static/mock/rss/hacker_news.xml");
        let result = parse(content).unwrap();
        assert_debug_snapshot!(result);
    }

    #[test]
    fn test_parse_rss_kotlin() {
        let content = include_str!("../../static/mock/rss/kotlin.xml");
        let result = parse(content).unwrap();
        assert_debug_snapshot!(result);
    }

    #[test]
    fn test_parse_rss_react() {
        let content = include_str!("../../static/mock/rss/react.xml");
        let result = parse(content).unwrap();
        assert_debug_snapshot!(result);
    }

    #[test]
    fn test_parse_rss_stopgame() {
        let content = include_str!("../../static/mock/rss/stopgame.xml");
        let result = parse(content).unwrap();
        assert_debug_snapshot!(result);
    }

    #[test]
    fn test_parse_rss_xkcd() {
        let content = include_str!("../../static/mock/rss/xkcd.xml");
        let result = parse(content).unwrap();
        assert_debug_snapshot!(result);
    }

    #[test]
    fn test_parse_date_handle_negative_zero_timezone() {
        let expected = chrono::DateTime::parse_from_rfc3339("2019-12-04T05:00:00+00:00").unwrap();
        assert_eq!(
            expected,
            parse_date("Wed, 04 Dec 2019 05:00:00 -0000").unwrap()
        );
    }
}
