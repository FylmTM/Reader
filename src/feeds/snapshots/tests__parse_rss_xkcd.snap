---
source: src/feeds/rss.rs
expression: result
---
[
    Post {
        id: -1,
        unique_id: "https://xkcd.com/2236/",
        link: "https://xkcd.com/2236/",
        title: "Is it Christmas?",
        date: 2019-12-02T05:00:00Z,
        summary: Some(
            "",
        ),
        content: Some(
            "<img src=\"https://imgs.xkcd.com/comics/is_it_christmas.png\" title=\"We\'ve tested it on 30 different days and it hasn\'t gotten one wrong yet.\" alt=\"We\'ve tested it on 30 different days and it hasn\'t gotten one wrong yet.\">",
        ),
        media_type: Some(
            MediaType {
                mime: "image/*",
            },
        ),
        media_link: Some(
            "https://imgs.xkcd.com/comics/is_it_christmas.png",
        ),
        comments_link: None,
    },
    Post {
        id: -1,
        unique_id: "https://xkcd.com/2237/",
        link: "https://xkcd.com/2237/",
        title: "AI Hiring Algorithm",
        date: 2019-12-04T05:00:00Z,
        summary: Some(
            "",
        ),
        content: Some(
            "<img src=\"https://imgs.xkcd.com/comics/ai_hiring_algorithm.png\" title=\"So glad Kate over in R&amp;D pushed for using the AlgoMaxAnalyzer to look into this. Hiring her was a great decisio- waaaait.\" alt=\"So glad Kate over in R&amp;D pushed for using the AlgoMaxAnalyzer to look into this. Hiring her was a great decisio- waaaait.\">",
        ),
        media_type: Some(
            MediaType {
                mime: "image/*",
            },
        ),
        media_link: Some(
            "https://imgs.xkcd.com/comics/ai_hiring_algorithm.png",
        ),
        comments_link: None,
    },
    Post {
        id: -1,
        unique_id: "https://xkcd.com/2238/",
        link: "https://xkcd.com/2238/",
        title: "Flu Shot",
        date: 2019-12-06T05:00:00Z,
        summary: Some(
            "",
        ),
        content: Some(
            "<img src=\"https://imgs.xkcd.com/comics/flu_shot.png\" title=\"&quot;Wait, how often are you getting bitten by snakes? And why are you boiling water?&quot; &quot;Dunno, the CDC people keep showing up with complicated questions about the \'history of the property\' and \'possible curses\' but I kinda tune them out. At least one of them offered me the flu shot.&quot;\" alt=\"&quot;Wait, how often are you getting bitten by snakes? And why are you boiling water?&quot; &quot;Dunno, the CDC people keep showing up with complicated questions about the \'history of the property\' and \'possible curses\' but I kinda tune them out. At least one of them offered me the flu shot.&quot;\">",
        ),
        media_type: Some(
            MediaType {
                mime: "image/*",
            },
        ),
        media_link: Some(
            "https://imgs.xkcd.com/comics/flu_shot.png",
        ),
        comments_link: None,
    },
    Post {
        id: -1,
        unique_id: "https://xkcd.com/2239/",
        link: "https://xkcd.com/2239/",
        title: "Data Error",
        date: 2019-12-09T05:00:00Z,
        summary: Some(
            "",
        ),
        content: Some(
            "<img src=\"https://imgs.xkcd.com/comics/data_error.png\" title=\"Cyanobacteria wiped out nearly all life on Earth once before, and they can do it again!\" alt=\"Cyanobacteria wiped out nearly all life on Earth once before, and they can do it again!\">",
        ),
        media_type: Some(
            MediaType {
                mime: "image/*",
            },
        ),
        media_link: Some(
            "https://imgs.xkcd.com/comics/data_error.png",
        ),
        comments_link: None,
    },
]
